be/src/storage/rowset/segment_rewriter.cpp

```diff
@@ -2,18 +2,116 @@
 
 SegmentRewriter::~SegmentRewriter() {}
 
+Status parse_segment_footer(std::unique_ptr<fs::ReadableBlock>* rblock, SegmentFooterPB* footer,
+                            size_t* footer_length) {
+    std::string buff;
+    raw::stl_string_resize_uninitialized(&buff, 12);
+
+    uint64_t file_size;
+    RETURN_IF_ERROR((*rblock)->size(&file_size));
+    RETURN_IF_ERROR((*rblock)->read(file_size - buff.size(), buff));
+
+    *footer_length = UNALIGNED_LOAD32(buff.data() + buff.size() - 12);
+    const uint32_t checksum = UNALIGNED_LOAD32(buff.data() + buff.size() - 8);
+    const uint32_t magic_number = UNALIGNED_LOAD32(buff.data() + buff.size() - 4);
+
+    if (magic_number != UNALIGNED_LOAD32(k_segment_magic)) {
+        return Status::Corruption(
+                strings::Substitute("Bad segment file $0: magic number not match", (*rblock)->path()));
+    }
+
+    if (file_size < 12 + *footer_length) {
+        return Status::Corruption(strings::Substitute("Bad segment file $0: file size $1 < $2", (*rblock)->path(),
+                                                      file_size, 12 + *footer_length));
+    }
+
+    raw::stl_string_resize_uninitialized(&buff, *footer_length);
+    RETURN_IF_ERROR((*rblock)->read(file_size - buff.size() - 12, buff));
+    uint32_t actual_checksum = crc32c::Value(buff.data(), buff.size());
+    if (actual_checksum != checksum) {
+        return Status::Corruption(
+                strings::Substitute("Bad segment file $0: footer checksum not match, actual=$1 vs expect=$2",
+                                    (*rblock)->path(), actual_checksum, checksum));
+    }
+
+    if (!footer->ParseFromArray(buff.data(), buff.size())) {
+        return Status::Corruption(
+                strings::Substitute("Bad segment file $0: failed to parse footer", (*rblock)->path()));
+    }
+
+    return Status::OK();
+}
+
 Status SegmentRewriter::rewrite(const std::string& src, const std::string& dest, const TabletSchema& tschema,
                                 std::vector<uint32_t>& column_ids,
-                                std::vector<std::unique_ptr<vectorized::Column>>& columns) {
-    // TODO: impl
+                                std::vector<std::unique_ptr<vectorized::Column>>& columns, size_t segment_id) {
+    fs::BlockManager* block_mgr = fs::fs_util::block_manager();
+    std::unique_ptr<fs::WritableBlock> wblock;
+    fs::CreateBlockOptions wblock_opts({dest});
+    RETURN_IF_ERROR(block_mgr->create_block(wblock_opts, &wblock));
+
+    std::unique_ptr<fs::ReadableBlock> rblock;
+    RETURN_IF_ERROR(block_mgr->open_block(src, &rblock));
+
+    uint64_t file_size;
+    RETURN_IF_ERROR(rblock->size(&file_size));
+
+    if (file_size < 12) {
+        return Status::Corruption(strings::Substitute("Bad segment file $0: file size $1 < 12", src, file_size));
+    }
+
+    SegmentFooterPB footer;
+    size_t footer_length = 0;
+    RETURN_IF_ERROR(parse_segment_footer(&rblock, &footer, &footer_length));
```

Comment:
How about moving this into parse_segment_footer, then returning a body length(as remaining)

Reply:
> How about moving this into parse_segment_footer, then returning a body length(as remaining)

done